builtin/rebase.c

```diff
@@ -0,0 +1,111 @@
+/*
+ * "git rebase" builtin command
+ *
+ * Copyright (c) 2018 Pratik Karki
+ */
+
+#include "builtin.h"
+#include "run-command.h"
+#include "exec-cmd.h"
+#include "parse-options.h"
+#include "argv-array.h"
+#include "refs.h"
+#include "dir.h"
+
+enum rebase_type {
+	PRESERVE_MERGES
+};
+
+static GIT_PATH_FUNC(git_path_rebase_apply, "rebase-apply");
+static GIT_PATH_FUNC(git_path_rebase_merge, "rebase-merge");
+
+enum rebase_type mode;
+
+static int use_builtin_rebase(void)
+{
```

Comment:
This whole function could be replaced by ```git_config_get_bool("rebase.usebuiltin", &ret)``` ? (I seem to be missing the need why we need to fork off a new process to load that boolean, as this is not part of a shell -> translation, but part of the setup of a wrapper, i.e. new code?)

Reply:
Yes. You're right. I will replace this function in the coming iterations as this has no more need. It was from the start as a placeholder.

Reply:
No, that's not right. The whole point of the spawning exercise is to avoid going through the `setup_git_directory()` dance too early (and *lose* the information about the current directory!)

Please study the commit message of the difftool commit that introduced this paradigm. If you deviate from the paradigm in any way, you will most likely introduce a bug.

Reply:
@dscho ok but then it would be nice if that was explained here in the commit message or in a comment (or both).

Reply:
Ok. @dscho  This seems to be the commit message.

> The current version of the builtin difftool does not, however, make full
> use of the internals but instead chooses to spawn a couple of Git
> processes, still, to make for an easier conversion. There remains a lot
> of room for improvement, left for a later date.

I'm sorry for not reading the message given by the commit properly. 
And I totally agree with you.
@chriscool I will add a comment for the WIP commit and for the patch I'll be adding this detail to the commit message.

Reply:
But that comment does not explain why you're avoiding calling git_config_get_bool(). It would be better to explain that you're avoiding the setup dance and why. See the initial commit adding builtin/difftool.c (be8a90e59ce) where there is a code comment explaining this.

Reply:
@dscho Why is it important to avoid changing directories when rebase doesn't have paths on the command line like difftool does?

Reply:
@phillipwood it is important in general, because code changes. So while the current code may not rely on the subdirectory in which it was started, future iterations might want to. To stay consistent is the sanest way to handle this: i.e. the way I suggested, to spawn.

In `rebase -i`'s case, you further have to take into account that there are potentially scripts making use of `GIT_PREFIX` (IIRC one of my scripts does, or at least did).

Reply:
FWIW there *is* one way you could avoid spawning, that was developed in the meantime: the early config machinery (use `read_early_config()`).

But the idea is to use this technique only for the time until the project is completed, so I would be totally cool with the simpler-to-reason-about spawning approach.